### PR TITLE
4. SecureStorage: read path

### PR DESCRIPTION
**Pull request by 0xB19**
- Add read path for decrypting and reading session data
- `decrypt_session_data_block` for individual block decryption
- `read_total_length` reads length header from block 0
- `read_session_data` provides pread-like interface with offset/length across block boundaries
- Checked arithmetic on all offset/length calculations
- 9 tests covering roundtrip, cross-block, offset, bounds, edge cases

### CODE REVIEW

File: wasm/bordercrypt/src/read.rs [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+pub fn decrypt_session_data_block<S: BlockStorage>(
+    storage: &S,
+    domain: &str,
+    session: &UnlockedSession,
+    block_index: u64,
+) -> Result<Zeroizing<[u8; PLAINTEXT_SIZE]>> {
+    if session.session_version != 0 {
+        return Err(BordercryptError::UnsupportedVersion(session.session_version));
+    }
+    let block_ct = storage.read_block(session.session_index, block_index)?;
```
Comment (Soulthym) [inline wasm/bordercrypt/src/read.rs]: just noticed storage.read_block matches read_session_block from the spec. I'd try to rename that, but probably in a followup so it doesnt cause issues while rebasing. I'm also not sure it's relevant (feel free to ignore if it is irrelevant).

File: wasm/bordercrypt/src/read.rs [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+        session.session_version,
+        session.session_index,
+        session.root_aead_key.as_ref(),
+        block_index,
+    );
+    decrypt_block(&session.pq_rerand_sk, &aead_key, &aad_root, &block_ct)
+}
+
+/// Read total data length from block 0 of an unlocked session.
+pub fn read_total_length<S: BlockStorage>(
```
Comment (Soulthym) [inline wasm/bordercrypt/src/read.rs:59]: this looks like a duplicate function name with this one:
https://github.com/massalabs/gossip/blob/324bcbc0de2ef79aaf922c89b1b2daed8aa8f953/wasm/secure-storage/src/unlock.rs#L119

Do those do the same thing? There docs seem to imply that. Maybe only use a single definition. The one in read.rs looks closer to the spec.

File: wasm/bordercrypt/src/read.rs [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+        .ok_or(BordercryptError::Overflow)?;
+    if end > session.total_data_length {
+        return Err(BordercryptError::OutOfBounds);
+    }
+
+    let ps = PLAINTEXT_SIZE as u64;
+    let hdr = LENGTH_HDR_SIZE as u64;
+
+    let start_pos = hdr.checked_add(offset).ok_or(BordercryptError::Overflow)?;
+    let end_pos = hdr.checked_add(end).ok_or(BordercryptError::Overflow)?;
```
Comment (Soulthym) [inline wasm/bordercrypt/src/read.rs]: Id rename this var to end_pos_excl like in the spec so it's clear why we later do a `- 1`.

File: wasm/bordercrypt/src/read.rs [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+        let plaintext = decrypt_session_data_block(storage, domain, session, block_idx)?;
+
+        let block_start = block_idx
+            .checked_mul(ps)
+            .ok_or(BordercryptError::Overflow)?;
+        let slice_start = if start_pos > block_start {
+            (start_pos - block_start) as usize
+        } else {
+            0
+        };
```
Comment (Soulthym) [inline wasm/bordercrypt/src/read.rs]: I would move that to after block_limit to match the spec.

Comment (Soulthym) [reply wasm/bordercrypt/src/read.rs]: I did check that its equivalent to the `max` line in read_session_data of the spec

File: wasm/bordercrypt/src/read.rs [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+        let block_limit = block_start
+            .checked_add(ps)
+            .ok_or(BordercryptError::Overflow)?;
+        let slice_end = if end_pos < block_limit {
+            (end_pos - block_start) as usize
+        } else {
+            PLAINTEXT_SIZE
+        };
```
Comment (Soulthym) [inline wasm/bordercrypt/src/read.rs]: I checked that its equivalent to the `min` line in read_session_data of the spec

**Review by Soulthym — CHANGES_REQUESTED** [review]
Left a couple comments below to improve the code quality and reduce duplication. The implementation looks correct according to the spec.

File: wasm/bordercrypt/src/read.rs (hunk 5 above)
Comment (0xB19) [reply wasm/bordercrypt/src/read.rs]: replaced by min

File: wasm/bordercrypt/src/read.rs (hunk 4 above)
Comment (0xB19) [reply wasm/bordercrypt/src/read.rs]: replaced

File: wasm/bordercrypt/src/read.rs (hunk 3 above)
Comment (0xB19) [reply wasm/bordercrypt/src/read.rs]: renamed thanks

File: wasm/bordercrypt/src/read.rs [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+    read_total_length_raw(
+        storage,
+        domain,
+        session.session_version,
+        session.session_index,
+        &session.pq_rerand_sk,
+        session.root_aead_key.as_ref(),
+    )
+}
+
```
Comment (Soulthym) [inline wasm/bordercrypt/src/read.rs]: I don't understand why we still need 2 functions doing the same thing? I would keep just the raw one, remove the read_total_length, then rename the raw one to `read_total_length`

File: wasm/bordercrypt/src/read.rs [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+    let mut result = Zeroizing::new(Vec::with_capacity(length));
+
+    for block_idx in first_block..=last_block {
+        let plaintext = decrypt_session_data_block(storage, domain, session, block_idx)?;
+
+        let block_start = block_idx
+            .checked_mul(ps)
+            .ok_or(BordercryptError::Overflow)?;
+        let slice_start = start_pos.saturating_sub(block_start) as usize;
+        let slice_end = ((end_pos_excl - block_start) as usize).min(PLAINTEXT_SIZE);
```
Comment (Soulthym) [inline wasm/bordercrypt/src/read.rs]: I realise my comment wasnt clear, that was correct algorithmically speaking.

Now it's somewhere between 2 versions, so if we're gonna rewrite that bit I'd stick to how the spec does it for consistency using min, max and block_start/end, ie:
```suggestion
        let block_start = block_idx
            .checked_mul(ps)
            .ok_or(secure-storageError::Overflow)?;
        let block_end = block_start
            .checked_add(PLAINTEXT_SIZE)
            .ok_or(secure-storageError::Overflow);

        // SAFETY: safe to substract without check since max guarantees the difference >= 0
        let slice_start = start_pos.max(block_start) - block_start;
        let slice_end = end_pos_excl
            .min(block_end)
            .checked_sub(block_start)
            .ok_or(secure-storageError::Overflow);
```

The current impl is harder to check for equivalence than the if/else blocks were, the suggestion above should restore 1-1 mapping to the spec

File: wasm/bordercrypt/src/unlock.rs [hunk 8]
```diff
@@ -90,14 +88,18 @@ pub fn unlock_session<S: BlockStorage + KeypairStorage>(
...
-                storage,
-                domain,
-                kf.version,
-                session,
-                &pq_rerand_sk,
-                &root_aead_key,
-            )?;
+            let total_data_length = if storage.block_count(session)? == 0 {
+                0
+            } else {
```
Comment (Soulthym) [inline wasm/bordercrypt/src/unlock.rs]: Is this if else really needed? it wasnt present before

**Review by Soulthym — CHANGES_REQUESTED** [review]
left a few comments below

File: wasm/bordercrypt/src/read.rs [hunk 9]
```diff
@@ -22,54 +22,43 @@ pub fn decrypt_session_data_block<S: BlockStorage>(
...
     storage: &S,
     domain: &str,
     version: u32,
     session_index: SessionIndex,
     pq_rerand_sk: &PqSecretKey,
     root_aead_key: &[u8],
 ) -> Result<u64> {
+    if storage.block_count(session_index)? == 0 {
+        return Ok(0);
+    }
```
Comment (Soulthym) [inline wasm/bordercrypt/src/read.rs:71]: I think this is an oversight in the spec itself. Doesnt hurt to check here anyway so looks good to me to keep that.
I'd add it to the spec under def read_total_length (section 9.2) : #380 

PS: this is only needed if we do not pre-allocate the first block for every session, I'm not exactly sure, may be worth adding a comment too to say why we check since it differs from the spec.

File: wasm/bordercrypt/src/read.rs [hunk 10]
```diff
@@ -0,0 +1,341 @@
...
+        &session.pq_rerand_sk,
+        session.root_aead_key.as_ref(),
+        block_index,
+    )
+}
+
+/// Read total data length by decrypting block 0.
+///
+/// Returns 0 if the session has no blocks yet.
+pub fn read_total_length<S: BlockStorage>(
```
Comment (Soulthym) [inline wasm/bordercrypt/src/read.rs:59]: Parts of it were identical to: https://github.com/massalabs/gossip/blob/83d1c116cb04a19224f4ef2be29c353bb69a0ba6/wasm/secure-storage/src/read.rs#L21-L32

Fixed in 30a04eacf112fd8632a952f0169b7806c0d560b7

**Review by Soulthym — CHANGES_REQUESTED** [review]
left 2 comments below, fixed one in 30a04ea and ran `cargo fmt` on the `bodercrypt` crate.

File: wasm/bordercrypt/src/read.rs (hunk 9 above)
Comment (Soulthym) [reply wasm/bordercrypt/src/read.rs:71]: fixed in 40f4ffb, also update the spec to add that check

**Comment by Soulthym** [comment]
fixed both issues in latest commits and updated the spec. lgtm!

**Review by Soulthym — APPROVED** [review]
